File: src/AlgoPlayground/Domain/ScoreEntry.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace AlgoPlayground.Domain;

public class ScoreEntry
{
    public int Id { get; set; }

    [Required]
    public GameId GameId { get; set; }

    [Range(1, 99)]
    public int Level { get; set; }

    [Range(0, int.MaxValue)]
    public int Points { get; set; }

    [Range(0, int.MaxValue)]
    public int DurationSec { get; set; }

    public DateTime Timestamp { get; set; } = DateTime.UtcNow;

    [ForeignKey(nameof(UserProfile))]
    public int UserProfileId { get; set; }

    public UserProfile? UserProfile { get; set; }
}

File: lib/core/models/score_entry.dart
import 'dart:convert';

import 'game_id.dart';

class ScoreEntry {
  const ScoreEntry({
    required this.gameId,
    required this.level,
    required this.points,
    required this.durationSec,
    required this.timestamp,
  });

  final GameId gameId;
  final int level;
  final int points;
  final int durationSec;
  final DateTime timestamp;

  Map<String, dynamic> toJson() => {
        'gameId': gameId.name,
        'level': level,
        'points': points,
        'durationSec': durationSec,
        'timestamp': timestamp.toIso8601String(),
      };

  factory ScoreEntry.fromJson(Map<String, dynamic> json) {
    return ScoreEntry(
      gameId: GameIdX.fromJson(json['gameId'] as String? ?? GameId.mazeRunner.name),
      level: json['level'] as int? ?? 1,
      points: json['points'] as int? ?? 0,
      durationSec: json['durationSec'] as int? ?? 0,
      timestamp: DateTime.tryParse(json['timestamp'] as String? ?? '') ?? DateTime.now(),
    );
  }

  static List<ScoreEntry> fromJsonString(String? jsonString) {
    if (jsonString == null || jsonString.isEmpty) {
      return [];
    }
    final dynamic data = jsonDecode(jsonString);
    if (data is List) {
      return data
          .whereType<Map<String, dynamic>>()
          .map(ScoreEntry.fromJson)
          .toList();
    }
    return [];
  }

  static String toJsonString(List<ScoreEntry> entries) {
    return jsonEncode(entries.map((e) => e.toJson()).toList());
  }
}
